File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> A;
int T, ans;

void dfs(int s, int sum) {
    if(s == A.size()) {
        if(sum == T) ans++;
        return;
    }
    
    dfs(s+1, sum+A[s]);
    dfs(s+1, sum-A[s]);
}

int solution(vector<int> numbers, int target) {
    A.assign(numbers.begin(), numbers.end());
    T = target;
    
    dfs(0, 0);
    
    return ans;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0

def dfs(numbers, target, idx, total):
    if idx == len(numbers):
        if total == target:
            global answer
            answer += 1
        return
    
    dfs(numbers, target, idx + 1, total + numbers[idx])
    dfs(numbers, target, idx + 1, total - numbers[idx])
        
    
def solution(numbers, target):
    dfs(numbers, target, 0, 0)
    return answer